### ISSUE TITLE
General interface for specify reference and positional indel scores

### ISSUE THREAD

**Issue by philiptzou** [hivdb/nucamino#3]

**Comment by neganp** [comment]
I was looking in to this and I've got a proposal that I'd like to run by you. It involves changing the CLI quite a bit.

Instead of having a separate command for each kind of alignment that `nucamino` can perform, I'd like to propose having a generic `align` command which takes an alignment profile as an argument. There will be some built-in profiles (such as the already-included HIV1b and HCV1a), or the user could specify one of their own. This syntax would look something like:

    $ nucamino align -g ns3 --profile hcv1a -i input.fasta -o output.tsv

To make the program more usable, there would need to be a command to list the available built-in alignment profiles.

    $ nucamino list-alignment-profiles
      Profile                Available Genes
      hcv1a                ns3 ns5a ns5b
      hiv1b                 gag pol gp41
    etc.

To help users write their own alignment profiles, there would need to be a command for checking alignment profiles to make sure they are formatted correctly and have all the required fields:

    $ nucamino check-alignment-profile < test-profile.json
      (ok or panic, dependig on whether the profile is valid)

With the user able to specify all the alignment parameters in a profile, we could deprecate the indel, gap, and codon penalty arguments (or keep them and have them over-ride the values in the selected profile, if that's more useful). Other arguments (e.g. goroutines, pprof) would be kept on the `align` command.

Both of the new commands are fairly straightforward to build, and the `align` command is mostly an adaptation of the existing code. I can work on it this week if this seems suitable.

Thoughts? Feedback? Is there a use case that I don't know about that this scheme wouldn't support?

**Comment by neganp** [comment]
Obviously, this would be full of breaking changes, and thus a major release of `nucamino`.

**Comment by philiptzou** [comment]
First of all, I like the idea of the generic `align` command and I think it is the one we should implement.

For the profile-related commands, the length of command is a concern for me. Instead of `list-alignment-profiles` and `check-alignment-profile`, I preferred `profile list` and `profile check`. Also the `-i test-profile.json` should be `--profile test-profile.json` to be consistent with the `align` command.

Furthermore, we should also discuss about the profile format and the built-in profiles. I preferred YAML format but I think it is doable to support both YAML and JSON formats.

**Comment by philiptzou** [comment]
Here is a proposal of the structure of alignment profiles. JSON format should use the same structure as YAML format. In `AAReference`, nucamino should automatically remove all space characters (spaces, tabs, newlines, etc) from the reference sequence.

```yaml
# hiv1b.yml
Globals:
  SubstitutionMatrix: @blosum62
  IndelCodonOpeningBonus: 0
  IndelCodonExtensionBonus: 2
  StopCodonPenalty: 4
  GapOpeningPenalty: 10
  GapExtensionPenalty: 2
Genes:
- GeneName: pol
  AAReference: >
    MGARASVLSGGELDRWEKIRLRPGGKKKYKLKHIVWASRELERFAVNPGLLETSEGCRQILGQLQPSLQT
    GSEELRSLYNTVATLYCVHQRIEVKDTKEALEKIEEEQNKSKKKAQQAAADTGNSSQVSQNYPIVQNLQG
    QMVHQAISPRTLNAWVKVVEEKAFSPEVIPMFSALSEGATPQDLNTMLNTVGGHQAAMQMLKETINEEAA
    EWDRLHPVHAGPIAPGQMREPRGSDIAGTTSTLQEQIGWMTNNPPIPVGEIYKRWIILGLNKIVRMYSPT
    SILDIRQGPKEPFRDYVDRFYKTLRAEQASQEVKNWMTETLLVQNANPDCKTILKALGPAATLEEMMTAC
    QGVGGPGHKARVLAEAMSQVTNSATIMMQRGNFRNQRKTVKCFNCGKEGHIAKNCRAPRKKGCWKCGKEG
    HQMKDCTERQANFLGKIWPSHKGRPGNFLQSRPEPTAPPEESFRFGEETTTPSQKQEPIDKELYPLASLR
    SLFGNDPSSQ
  PositionalIndelCodonScores:
  # type, position, open, ext
  - [ins, 218, -5, 0]  # RT63
  - [del, 218, -5, 0]
  - [ins, 219, -5, 0]  # RT64
  - [del, 219, -5, 0]
  - [ins, 220, -7, 0]  # RT65
  - [ins, 221, -7, 0]  # RT66
  - [ins, 222, -7, 0]  # RT67
  - [ins, 223, -3, 0]  # RT68
  - [del, 223, 0, 0]
  - [ins, 224, 18, -3]  # RT69
  - [ins, 225, -3, 0]  # RT70
  - [ins, 226, -3, 0]  # RT71
  - [ins, 227, -3, 0]  # RT72
  - [ins, 228, -3, 0]  # RT73
- GeneName: gag
  AAReference: ...

```

The built-in profiles should use the same structure as above defined. However, we need to write a pre-compiler script to convert the built-in profiles into `.go` files.

**Comment by neganp** [comment]
Regarding the CLI, I can get behind those changes. What do you think of taking the refactoring as an opportunity to migrate from go-flags to the better-supported [cobra](https://github.com/spf13/cobra)?

Regarding the profile format, I was just going to use the JSON/YAML representation of the Profile struct, which would look something like

```json
{
    "StopCodonPenalty": 0,
    "GapOpeningPenalty": 1,
    "GapExtensionPenalty": 2,
    "IndelCodonOpeningBonus": 3,
    "IndelCodonExtensionBonus": 4,
    "ReferenceSequences": {
        "A": "MLNGIKLENQEAMDWTHFYQDNEVYSGV"
    },
    "GeneIndelScores": {
        "A": {"1": [2, 3], "-4": [5, 6]}
    }
}
```

This makes the saving and loading code very straightforward; I've implemented it on a branch. I do like the indel codon scores format above. It's more explicit than the straight conversion, but it's that much more code to support it.

As a side-effect of the way the JSON parsing works, extra keys (for things like comments, genbank accession ids, etc.) can be added to a profile and they won't show up when it's parsed.

I also prefer YAML, if only for the ability to add comments, but I wasn't sure which of the Golang YAML parsers are current, supported, and/or worth using, so I held off for now. Do you have a preference?

**Comment by neganp** [comment]
I would also agree that the built-in profiles should use the same format. They could even just be embedded as strings and parsed at startup time; the overhead is pretty minimal.

**Comment by philiptzou** [comment]
You can go ahead to replace go-flags with cobra. I agreed we should use a better supported library.

About the format, I preferred a structure that allows per-gene penalty/bonus scores. I could write some code to convert the explicit indel codon scores format into the internal format. Therefore we don't have to change the core.

For YAML format, I think the famous one is [go-yaml/yaml](https://github.com/go-yaml/yaml). The latest update is two months ago so I think it is current, right?

**Comment by philiptzou** [comment]
> I would also agree that the built-in profiles should use the same format. They could even just be embedded as strings and parsed at startup time; the overhead is pretty minimal.

We could save some time when calling nucamino if we can pre-parse the built-in profiles (that's why I said we can convert the profiles into `.go` files).

**Comment by philiptzou** [comment]
@neganp Sorry for having you to do all the coding works so far. I am appreciated for your contributions. Currently I am busy in another project of ours (which will be open source eventually). But I'd like to provide my suggestions and probably find some time to help on the coding.

**Comment by neganp** [comment]
Thanks for the library recommendation! If we support YAML, I don't think we also have to support JSON. I'll get to work on converting my branch to use go-yaml/yaml and cobra.

I'll also work on changing the CLI interface to have the commands:

- align
- profile list
- profile check

with the arguments we discussed above.

Regarding the format, I think it will be much easier to decode if we keep it closer to struct it will eventually be, so something like:

```yaml
IndelCodonOpeningBonus: 0
IndelCodonExtensionBonus: 2
StopCodonPenalty: 4
GapOpeningPenalty: 10
GapExtensionPenalty: 2
ReferenceSequences:
  pol: >
    MGARASVLSGGELDRWEKIRLRPGGKKKYKLKHIVWASRELERFAVNPGLLETSEGCRQILGQLQPSLQT
    GSEELRSLYNTVATLYCVHQRIEVKDTKEALEKIEEEQNKSKKKAQQAAADTGNSSQVSQNYPIVQNLQG
    QMVHQAISPRTLNAWVKVVEEKAFSPEVIPMFSALSEGATPQDLNTMLNTVGGHQAAMQMLKETINEEAA
    EWDRLHPVHAGPIAPGQMREPRGSDIAGTTSTLQEQIGWMTNNPPIPVGEIYKRWIILGLNKIVRMYSPT
    SILDIRQGPKEPFRDYVDRFYKTLRAEQASQEVKNWMTETLLVQNANPDCKTILKALGPAATLEEMMTAC
    QGVGGPGHKARVLAEAMSQVTNSATIMMQRGNFRNQRKTVKCFNCGKEGHIAKNCRAPRKKGCWKCGKEG
    HQMKDCTERQANFLGKIWPSHKGRPGNFLQSRPEPTAPPEESFRFGEETTTPSQKQEPIDKELYPLASLR
    SLFGNDPSSQ
  gag: >
    etc.
  gp41:
    etc.

PositionalIndelCodonScores:
  pol:
    # type, position, open, ext
    - [ins, 218, -5, 0]  # RT63
    - [del, 218, -5, 0]
    - [ins, 219, -5, 0]  # RT64
    - [del, 219, -5, 0]
    - [ins, 220, -7, 0]  # RT65
    - [ins, 221, -7, 0]  # RT66
    - [ins, 222, -7, 0]  # RT67
    - [ins, 223, -3, 0]  # RT68
    - [del, 223, 0, 0]
    - [ins, 224, 18, -3]  # RT69
    - [ins, 225, -3, 0]  # RT70
    - [ins, 226, -3, 0]  # RT71
    - [ins, 227, -3, 0]  # RT72
    - [ins, 228, -3, 0]  # RT73
  gag:
    - etc.
  gp41:
    - etc.
```

Reference sequences live in a map of gene name to string, which gets parsed into array of AminoAcids. Positional indel scores live in a different (optional) map of gene names which gets parsed into the indel score structure that the core algorithm expects.

Does that look reasonable?

**Comment by philiptzou** [comment]
@neganp It looks good to me as the first version of alignment profile. I'd like to introduce the substitution matrix into the profile but I think we can skip it in the first version.

**Comment by neganp** [comment]
:+1: Sounds good. I think that will also require some changes to the core code (as Blosum is currently hard-coded), and this change is sprawling enough already, so if we can open an issue for it and return to it later, I think it will make this more manageable.

Thanks for your feedback! I'll get to work on this.

**Comment by philiptzou** [comment]
@neganp Thanks!

**Comment by neganp** [comment]
I've run in to some trouble getting the YAML serialization of the alignment profile to behave. Specifically, 
I don't think [go-yaml](https://github.com/go-yaml/yaml) lets me specify how to format an array-of-arrays, which is how we want to set up the positional indel scores. 

It can be configured to do this, with the arrays all laid out long-form:

```yaml
StopCodonPenalty: 0
GapOpeningPenalty: 1
GapExtensionPenalty: 2
IndelCodonOpeningBonus: 3
IndelCodonExtensionBonus: 4
GeneIndelScore:
  A:
  - - ins
    - 1
    - 2
    - 3
  - - del
    - 4
    - 5
    - 6
  B:
  - - ins
    - 5
    - 2
    - 1
ReferenceSequences:
  A: DWTHFYQDNEVYSGV
  B: MLNGIKLENQEAM
```

Or like this, with the arrays all nested: 

```yaml
StopCodonPenalty: 0
GapOpeningPenalty: 1
GapExtensionPenalty: 2
IndelCodonOpeningBonus: 3
IndelCodonExtensionBonus: 4
GeneIndelScore: {A: [[ins, 1, 2, 3], [del, 4, 5, 6]], B: [[ins, 5, 2, 1]]}
ReferenceSequences:
  A: DWTHFYQDNEVYSGV
  B: MLNGIKLENQEAM
```

I can't figure out how to configure it to split the different and lay out the inner arrays in-line and the outer array with bullet points. The root of the problem is that the package uses struct field annotations to tell the encoder which style to use. That options isn't availble on an array of arrays (as there's no struct field to put the annotations on) and the encoder doesn't expose an API to do it a different way.

For this reason, I think I'll use the built-in `template` package to serialize the profile rather than the yaml package. Parsing seems to work fine either way, so that's not a problem.

**Comment by philiptzou** [comment]
I forgot that we have the built-in `template` in go. Yes we definitely should consider using it. Thanks for bringing this up.

**Comment by neganp** [comment]
I think #10 should wrap up this issue.

**Comment by philiptzou** [comment]
Solved and 0.2.0a1 released.
